src/plugins/hetero/src/compiled_model.cpp

```diff
@@ -0,0 +1,708 @@
+            for (auto&& node : orderedOps) {
+                auto& nodeSubgraphInputDependency = nodeSubgraphInputDependencies[node];
+                auto allNodeSubgraphInputs =
+                    InferenceEngine::details::Intersection(nodeInputDependencies[node], subgraphInputs);
+                for (auto&& subgraphInput : allNodeSubgraphInputs) {
+                    if (subgraphIds[node] == subgraphIds[subgraphInput.get_node()->shared_from_this()]) {
+                        nodeSubgraphInputDependency.emplace(subgraphInput);
+                    }
+                }
+                auto& nodeSubgraphCyclicInputDependency = nodeSubgraphCyclicInputDependencies[node];
+                for (auto&& subgraphInput : allNodeSubgraphInputs) {
+                    if (!ov::op::util::is_parameter(subgraphInput.get_node()) &&
+                        !ov::op::util::is_constant(subgraphInput.get_node()) &&
+                        subgraphIds[node] == subgraphIds[InputNode(subgraphInput)]) {
+                        nodeSubgraphCyclicInputDependency.emplace(subgraphInput);
+                    }
+                }
+            }
+
+            for (auto&& node : orderedOps) {
+                auto& nodeSubgraphCyclicInputDependency = nodeSubgraphCyclicInputDependencies[node];
+                if (!nodeSubgraphCyclicInputDependency.empty()) {
+                    // Collect all subgraph inputs that cyclic subgraph output depends on
+                    InputSet cyclicInputsDependencies;
+                    for (auto&& cyclicInput : nodeSubgraphCyclicInputDependency) {
+                        for (auto&& input : nodeSubgraphInputDependencies[InputNode(cyclicInput)]) {
+                            cyclicInputsDependencies.emplace(input);
+                        }
+                    }
+                    for (auto&& input : node->inputs()) {
+                        auto& inputNodeSubgraphCyclicInputDependency =
+                            nodeSubgraphCyclicInputDependencies[InputNode(input)];
+                        auto& inputNodeSubgraphInputDependency = nodeSubgraphInputDependencies[InputNode(input)];
+                        if (!InferenceEngine::details::Intersects(nodeSubgraphCyclicInputDependency,
+                                                                  inputNodeSubgraphCyclicInputDependency) &&
+                            InferenceEngine::details::Intersects(cyclicInputsDependencies,
+                                                                 inputNodeSubgraphInputDependency)) {
+                            subgraphInputs.insert(input);
+                        }
+                    }
+                }
+            }
+        }
+
+        auto subgraphIds = CollectSubgraphs();
+
+        if (dumpDotFile) {
+            std::map<std::string, int> map_id;
+            for (auto&& v : subgraphIds) {
+                map_id.emplace(v.first->get_friendly_name(), v.second);
+            }
```

Comment:
Can it be a part of dump subgraphs method?

Reply:
For the moment I see no good alternative. `subgraphIds` has overcomplicated type for `dump_affinities()` to be passed as is